doc/module-system: Introduce concepts (#390993)
NixOS/nixpkgs: doc/module-system/module-system.chapter.md

# Module System {#module-system}

::: {.note}
This chapter is new and not complete yet. For a gentle introduction to the module system, in the context of NixOS, see [Writing NixOS Modules](https://nixos.org/manual/nixos/unstable/index.html#sec-writing-modules) in the NixOS manual.
:::

## Introduction {#module-system-introduction}

::: {.note}
This chapter is new and not complete yet.
:::

The module system is a language for handling configuration, implemented as a Nix library.

> It was historically known as the NixOS module system.

Reviewer: ```suggestion
It was historically known as the NixOS module system, but the mechanism is not specific to NixOS.
```

Its purpose is to "extend" the Nix language with

> - **dynamic type checking** to find mistakes early on, improving error messages;

Reviewer: ```suggestion
- **dynamic type checking** to catch and locate mistakes early on;
```

It would be a bit tone deaf to speak about improving error messages from the position of official documentation.

- **documentation**, so that each option has a description, conveniently located in the file that declares it
- **modularity** or **aspect oriented programming**: separate files can all contribute to the same option.

> The following sections are dense introductory reference material.

Reviewer: ```suggestion
The following sections are introductory reference material.
```

Not sure what "dense" would mean.
